### ISSUE TITLE
Fake tensor support of distributed ops in FX graph tracing

### ISSUE THREAD

**Issue by xuzijian629** [pytorch/pytorch#128128]
### 🚀 The feature, motivation and pitch

Currently, the FX graph tracing (such as the one used in `aot_module`) seems not supporting collective functions such as `allgather`.

We may face some errors like `torch._subclasses.fake_tensor.UnsupportedOperatorException: c10d._allgather_base_.default` when tracing modules containing these operators.

Repro example:

<details>

```bash
 mpiexec -N 2 python3 repro.py
```

```python3
import torch
import os
from torch._functorch.aot_autograd import aot_module


rank = int(os.environ["OMPI_COMM_WORLD_RANK"])
world_size = int(os.environ["OMPI_COMM_WORLD_SIZE"])
torch.distributed.init_process_group(
    init_method="file:///tmp/dist_init",
    rank=rank,
    world_size=world_size,
    backend="gloo",
)


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        pg = torch.distributed.group.WORLD
        out = torch.empty_like(x).expand(world_size, -1).reshape(-1)
        work = pg._allgather_base(out, x)
        work.wait()
        return out


model = Model()
x = torch.rand(10)

model = aot_module(model, fw_compiler=None)

model(x)
```

</details>

Supporting the trace for these operators enables optimizing the computational graphs in DistributedDataParallel training.
In DistributedDataParallel, one of the common technique might be

1. First wrap the model by aot_module
2. Then, wrap the model in 1 by DistributedDataParallel

so that the allreduce in the gradient computation will not be traced in the FX graph.

However, this cannot be eliminate the errors when `SyncBatchNorm`, which syncs running stats among GPUs, for example.
To support these cases, we need the abstract implemetation of the distributed ops.

### Alternatives

Some standard idea could be adding "Meta" dispatch key for `c10d._allgather_base_` manually to support its tracing by fake tensors.

But this needs the modification of C++ source files to add abstract impl pystub.

Another temporal workaround could be defining a custom allgather via torch.library with `Meta` dispatcher supported.

However, in most practical cases, `c10d._allgather_base_` is implicitly called by torch internals or other frameworks (such as pytorch_lightning). In such situations, replacing the function could be difficult.

**Comment by awgu** [comment]
It might be worth taking a look at the discussion in https://github.com/pytorch/pytorch/issues/93173.

**Comment by yifuwang** [comment]
We should probably migrate things like `SyncBatchNorm` to use functional collectives.

**Comment by xuzijian629** [comment]
Thanks for introducing functional collectives to me! It seems they will resolve this problem.
Please close this issue if you think having #93173 is enough.
